# Web Scrapping Project - We will be scraping from Wikipedia

### First we need to load all the necessary libraries:

In [288]:
# Importing the libraries
import pandas as pd #library to work with data frames
import numpy as np #library to work with arrays
from bs4 import BeautifulSoup as bs #library to scrape from website
import requests #library to make requests to website.
import requests # library to handle requests
from pandas.io.json import json_normalize
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

### Then we are requesting the http to get the text of the website

In [289]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

### Now all we have to do is using BeautifulSoup to load the data and build the pandas dataframe.

In [290]:
soup = bs(source, 'lxml')
soup.prettify()
header=[]
htemp=soup.find("table").tr

#Getting the header into a list
counter=0
for x in htemp.find_all("th"):
    header.append(x.text)
    
#Getting the body into a list with different columns
bod=soup.find("table")
body=[]
for i in bod.find_all("td"):
    body.append(i.text)
    #print(i.text)
    
Postcode = body[0:len(body):3]
Borough = body[1:len(body):3]
Neighborhood = body[2:len(body):3]
Neighborhood= [s.replace('\n', '') for s in Neighborhood] #Dropping the \n

#Creating Pandas dataframe
df=pd.DataFrame({'Borough':Borough,
                 'Postcode':Postcode,
                 'Neighborhood':Neighborhood})

#Reorganizing the dataframe's columns
df=df[['Postcode','Borough','Neighborhood']]

#Creating index list with 'Not assigned' boroughs
emptybor=df.index[df['Borough']=='Not assigned'].tolist()

#Dropping the emptyboroughs
df=df.drop(emptybor,axis=0)

#reseting the indexes of the new dataframe
df=df.reset_index(drop=True)

# Finding the indexes for "Not assigned" neighborhoods:
nan=df.index[df['Neighborhood']=='Not assigned'].tolist()

#modifying the "Not assigned" neighborhoods to Boroughs
df.iloc[nan,2] =df.iloc[nan,1]

# Using groupby we can obtain the list of neighborhoods that are in same postcode.
df = df.groupby(['Postcode','Borough'])['Neighborhood'].apply(list)

# Reseting the indexes
df=df.reset_index()

#looking at the first 5 items of final dataframe
df


Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A  Downtown Toronto   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                          Neighborhood  
0                                     [Rouge, Malvern]  
1             [Highland Creek, Rouge Hill, Port Union]  
2                  [Guildwood, Morningside, West Hill]  
3                                             [Woburn]  
4                                          [Cedarbrae]  
5                                [Scarborough Village]  
6        [East Birchmount Park, Ionview, Kennedy Park]  
7                    [Clairlea, Golden Mile, Oakridge]  
8    [Cliffcrest, Cliffside, Scarborough Village West]  
9                        [Birch Cliff, Cliffside West]  
10   [Dorset Park, Scarborough Town Centre, Wexford...  
11                                 [Maryvale, Wexford]  
12                                         [Agincourt]  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  
15                                   [L'Amoreaux West]  
16                                       [Upper Rouge]  
17                                 [Hillcrest Village]  
18                      [Fairview, Henry Farm, Oriole]  
19                                   [Bayview Village]  
20                          [Silver Hills, York Mills]  
21                           [Newtonbrook, Willowdale]  
22                                  [Willowdale South]  
23                                   [York Mills West]  
24                                   [Willowdale West]  
25                                         [Parkwoods]  
26                                   [Don Mills North]  
27                  [Flemingdon Park, Don Mills South]  
28   [Bathurst Manor, Downsview North, Wilson Heights]  
29                   [Northwood Park, York University]  
..                                                 ...  
73                                [Humewood-Cedarvale]  
74            

In [291]:
# Lets see whats the shape of our final dataframe:
df.shape

(103, 3)

The final dataframe contains 103 rows

## Now we are going to find the lattitude and longitude based on our post_codes:

In [292]:
import sys
!{sys.executable} -m pip install geocoder


You are using pip version 18.1, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [293]:
import geocoder # import geocoder

# initializing the variable to None
lat_lng_coords = None

# loop until we get the coordinates
while(lat_lng_coords is None):
    g = geocoder.google('Mountain View, CA', sensor=True)
    lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]
latitude

KeyboardInterrupt: 

## geocoder did not quite work. Therefore, below I will use the .csv file to add lattitude and longitude data

In [294]:
import os
os.getcwd()


'/Users/John/Desktop/Data Scientist package/IBM Coursera Data analytics/Course 9/Data Scraping project'

In [295]:
latlong=pd.read_csv('Geospatial_Coordinates.csv')
lat=latlong['Latitude']
lng=latlong['Longitude']
df['Latitude']=lat
df['Longitude']=lng
df


Postcode           Borough  \
0        M1B       Scarborough   
1        M1C       Scarborough   
2        M1E       Scarborough   
3        M1G       Scarborough   
4        M1H       Scarborough   
5        M1J       Scarborough   
6        M1K       Scarborough   
7        M1L       Scarborough   
8        M1M       Scarborough   
9        M1N       Scarborough   
10       M1P       Scarborough   
11       M1R       Scarborough   
12       M1S       Scarborough   
13       M1T       Scarborough   
14       M1V       Scarborough   
15       M1W       Scarborough   
16       M1X       Scarborough   
17       M2H        North York   
18       M2J        North York   
19       M2K        North York   
20       M2L        North York   
21       M2M        North York   
22       M2N        North York   
23       M2P        North York   
24       M2R        North York   
25       M3A        North York   
26       M3B        North York   
27       M3C        North York   
28       M3H        North York   
29       M3J        North York   
..       ...               ...   
73       M6C              York   
74       M6E              York   
75       M6G  Downtown Toronto   
76       M6H      West Toronto   
77       M6J      West Toronto   
78       M6K      West Toronto   
79       M6L        North York   
80       M6M              York   
81       M6N              York   
82       M6P      West Toronto   
83       M6R      West Toronto   
84       M6S      West Toronto   
85       M7A      Queen's Park   
86       M7R       Mississauga   
87       M7Y      East Toronto   
88       M8V         Etobicoke   
89       M8W         Etobicoke   
90       M8X         Etobicoke   
91       M8Y         Etobicoke   
92       M8Z         Etobicoke   
93       M9A  Downtown Toronto   
94       M9B         Etobicoke   
95       M9C         Etobicoke   
96       M9L        North York   
97       M9M        North York   
98       M9N              York   
99       M9P         Etobicoke   
100      M9R         Etobicoke   
101      M9V         Etobicoke   
102      M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                     [Rouge, Malvern]  43.806686 -79.194353  
1             [Highland Creek, Rouge Hill, Port Union]  43.784535 -79.160497  
2                  [Guildwood, Morningside, West Hill]  43.763573 -79.188711  
3                                             [Woburn]  43.770992 -79.216917  
4                                          [Cedarbrae]  43.773136 -79.239476  
5                                [Scarborough Village]  43.744734 -79.239476  
6        [East Birchmount Park, Ionview, Kennedy Park]  43.727929 -79.262029  
7                    [Clairlea, Golden Mile, Oakridge]  43.711112 -79.284577  
8    [Cliffcrest, Cliffside, Scarborough Village West]  43.716316 -79.239476  
9                        [Birch Cliff, Cliffside West]  43.692657 -79.264848  
10   [Dorset Park, Scarborough Town Centre, Wexford...  43.757410 -79.273304  
11                                 [Maryvale, Wexford]  43.750072 -79.295849  
12                                         [Agincourt]  43.794200 -79.262029  
13           [Clarks Corners, Sullivan, Tam O'Shanter]  43.781638 -79.304302  
14   [Agincourt North, L'Amoreaux East, Milliken, S...  43.815252 -79.284577  
15                                   [L'Amoreaux West]  43.799525 -79.318389  
16                                       [Upper Rouge]  43.836125 -79.205636  
17                                 [Hillcrest Village]  43.803762 -79.363452  
18                      [Fairview, Henry Farm, Oriole]  43.778517 -79.346556  
19                                   [Bayview Village]  43.786947 -79.385975  
20                          [Silver Hills, York Mills]  43.757490 -79.374714  
21                           [Newtonbrook, Willowdale]  43.789053 -79.408493  
22                                  [Willowdale South]  43.770120 -79.408493  

## Expanding neighborhood to single rows:

In [296]:
def expand_list(df, list_column, new_column): 
    lens_of_lists = df[list_column].apply(len)
    origin_rows = range(df.shape[0])
    destination_rows = np.repeat(origin_rows, lens_of_lists)
    non_list_cols = (
      [idx for idx, col in enumerate(df.columns)
       if col != list_column]
    )
    expanded_df = df.iloc[destination_rows, non_list_cols].copy()
    expanded_df[new_column] = (
      [item for items in df[list_column] for item in items]
      )
    expanded_df.reset_index(inplace=True, drop=True)
    return expanded_df

#usage
expanded_df = expand_list(df,"Neighborhood", "Neighborhood")
cols=expanded_df.columns.tolist()
cols=cols[0:2]+cols[-1:]+cols[2:4]
expanded_df=expanded_df[cols]
expanded_df.head()

Postcode      Borough    Neighborhood   Latitude  Longitude
0      M1B  Scarborough           Rouge  43.806686 -79.194353
1      M1B  Scarborough         Malvern  43.806686 -79.194353
2      M1C  Scarborough  Highland Creek  43.784535 -79.160497
3      M1C  Scarborough      Rouge Hill  43.784535 -79.160497
4      M1C  Scarborough      Port Union  43.784535 -79.160497

## I will filter only cities that has 'Toronto' in it

In [297]:
expanded_df.head()

Postcode      Borough    Neighborhood   Latitude  Longitude
0      M1B  Scarborough           Rouge  43.806686 -79.194353
1      M1B  Scarborough         Malvern  43.806686 -79.194353
2      M1C  Scarborough  Highland Creek  43.784535 -79.160497
3      M1C  Scarborough      Rouge Hill  43.784535 -79.160497
4      M1C  Scarborough      Port Union  43.784535 -79.160497

In [298]:
df_toronly=expanded_df[expanded_df['Borough'].str.contains('Toronto')]

In [299]:
df_toronly.reset_index(drop=True, inplace=True)
df_toronly


Postcode           Borough  \
0       M4E      East Toronto   
1       M4K      East Toronto   
2       M4K      East Toronto   
3       M4L      East Toronto   
4       M4L      East Toronto   
5       M4M      East Toronto   
6       M4N   Central Toronto   
7       M4P   Central Toronto   
8       M4R   Central Toronto   
9       M4S   Central Toronto   
10      M4T   Central Toronto   
11      M4T   Central Toronto   
12      M4V   Central Toronto   
13      M4V   Central Toronto   
14      M4V   Central Toronto   
15      M4V   Central Toronto   
16      M4V   Central Toronto   
17      M4W  Downtown Toronto   
18      M4X  Downtown Toronto   
19      M4X  Downtown Toronto   
20      M4Y  Downtown Toronto   
21      M5A  Downtown Toronto   
22      M5B  Downtown Toronto   
23      M5B  Downtown Toronto   
24      M5C  Downtown Toronto   
25      M5E  Downtown Toronto   
26      M5G  Downtown Toronto   
27      M5H  Downtown Toronto   
28      M5H  Downtown Toronto   
29      M5H  Downtown Toronto   
..      ...               ...   
44      M5S  Downtown Toronto   
45      M5T  Downtown Toronto   
46      M5T  Downtown Toronto   
47      M5T  Downtown Toronto   
48      M5V  Downtown Toronto   
49      M5V  Downtown Toronto   
50      M5V  Downtown Toronto   
51      M5V  Downtown Toronto   
52      M5V  Downtown Toronto   
53      M5V  Downtown Toronto   
54      M5V  Downtown Toronto   
55      M5W  Downtown Toronto   
56      M5X  Downtown Toronto   
57      M5X  Downtown Toronto   
58      M6G  Downtown Toronto   
59      M6H      West Toronto   
60      M6H      West Toronto   
61      M6J      West Toronto   
62      M6J      West Toronto   
63      M6K      West Toronto   
64      M6K      West Toronto   
65      M6K      West Toronto   
66      M6P      West Toronto   
67      M6P      West Toronto   
68      M6R      West Toronto   
69      M6R      West Toronto   
70      M6S      West Toronto   
71      M6S      West Toronto   
72      M7Y      East Toronto   
73      M9A  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
0                                         The Beaches  43.676357 -79.293031  
1                                   The Danforth West  43.679557 -79.352188  
2                                           Riverdale  43.679557 -79.352188  
3                                    The Beaches West  43.668999 -79.315572  
4                                        India Bazaar  43.668999 -79.315572  
5                                     Studio District  43.659526 -79.340923  
6                                       Lawrence Park  43.728020 -79.388790  
7                                    Davisville North  43.712751 -79.390197  
8                                  North Toronto West  43.715383 -79.405678  
9                                          Davisville  43.704324 -79.388790  
10                                         Moore Park  43.689574 -79.383160  
11                                    Summerhill East  43.689574 -79.383160  
12                                          Deer Park  43.686412 -79.400049  
13                                     Forest Hill SE  43.686412 -79.400049  
14                                          Rathnelly  43.686412 -79.400049  
15                                         South Hill  43.686412 -79.400049  
16                                    Summerhill West  43.686412 -79.400049  
17                                           Rosedale  43.679563 -79.377529  
18                                        Cabbagetown  43.667967 -79.367675  
19                                     St. James Town  43.667967 -79.367675  
20                               Church and Wellesley  43.665860 -79.383160  
21                                       Harbourfront  43.654260 -79.360636  
22                                            Ryerson  43.657162 -79.378937  
23                                    Garden District  43.657162 -79.378937  
24    

## Getting Geographical location (latitude and longitude of Toronto)

In [300]:
#Importing Nominatim library to get the location data of an address
from geopy.geocoders import Nominatim


In [301]:
address = 'Toronto'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


## Visualizing the map of Toronto with the neighbourhoods

In [302]:
#importing library for maps
import folium

In [303]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df_toronly['Latitude'], df_toronly['Longitude'], df_toronly['Neighborhood']):
    label = folium.Popup(str(label), parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Defining Foursquare credential to start exploring and segmenting the neighbourhoods

In [304]:
CLIENT_ID = '4Q2A2DMEZWFE0TMMPLCPZDCOUW0LKU2CTDDVMSJ5P2LIPPCC' # your Foursquare ID
CLIENT_SECRET = 'GTTVOIPQWRHLRZNDXUKZO52SVKNM0F4J0VU5QSFQV0LIOAHD' # your Foursquare Secret
VERSION = '20190911' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4Q2A2DMEZWFE0TMMPLCPZDCOUW0LKU2CTDDVMSJ5P2LIPPCC
CLIENT_SECRET:GTTVOIPQWRHLRZNDXUKZO52SVKNM0F4J0VU5QSFQV0LIOAHD


## Creating function to get all venues from neighbourhoods of Toronto

In [305]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [306]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


# Getting the toronto venues:
toronto_venues = getNearbyVenues(names= df_toronly['Neighborhood'],
                                   latitudes= df_toronly['Latitude'],
                                   longitudes= df_toronly['Longitude']
                                  )

The Beaches
The Danforth West
Riverdale
The Beaches West
India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park
Summerhill East
Deer Park
Forest Hill SE
Rathnelly
South Hill
Summerhill West
Rosedale
Cabbagetown
St. James Town
Church and Wellesley
Harbourfront
Ryerson
Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide
King
Richmond
Harbourfront East
Toronto Islands
Union Station
Design Exchange
Toronto Dominion Centre
Commerce Court
Victoria Hotel
Roselawn
Forest Hill North
Forest Hill West
The Annex
North Midtown
Yorkville
Harbord
University of Toronto
Chinatown
Grange Park
Kensington Market
CN Tower
Bathurst Quay
Island airport
Harbourfront West
King and Spadina
Railway Lands
South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place
Underground city
Christie
Dovercourt Village
Dufferin
Little Portugal
Trinity
Brockton
Exhibition Place
Parkdale Village
High Park
The Junction South
Parkdale
Roncesvalles
Runnymede

In [307]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                 Glen Stewart Ravine       43.676300       -79.294784   
4                       Upper Beaches       43.680563       -79.292869   

         Venue Category  
0                 Trail  
1     Health Food Store  
2                   Pub  
3  Other Great Outdoors  
4          Neighborhood

## Lets see how many venues were returned for each neighbourhood

In [308]:
toronto_venues.groupby('Neighborhood').count().sort_values(by='Venue',ascending=False)

Neighborhood Latitude  \
Neighborhood                                                               
St. James Town                                                       143   
Adelaide                                                             100   
Garden District                                                      100   
Harbourfront East                                                    100   
First Canadian Place                                                 100   
King                                                                 100   
Richmond                                                             100   
Design Exchange                                                      100   
Ryerson                                                              100   
Commerce Court                                                       100   
Toronto Dominion Centre                                              100   
Toronto Islands                                                      100   
Underground city                                                     100   
Union Station                                                        100   
Victoria Hotel                                                       100   
Stn A PO Boxes 25 The Esplanade                                       96   
Chinatown                                                             94   
Kensington Market                                                     94   
Grange Park                                                           94   
Central Bay Street                                                    84   
Church and Wellesley                                                  83   
Trinity                                                               57   
Little Portugal                                                       57   
Berczy Park                                                           55   
Harbourfront                                                          50   
Cabbagetown                                                           43   
The Danforth West                                                     41   
Riverdale                                                             41   
Harbord                                                               38   
Studio District                                                       38   
...                                                                  ...   
Yorkville                                                             20   
North Midtown                                                         20   
The Annex                                                             20   
Dovercourt Village                                                    17   
Christie                                                              17   
Dufferin                                                              17   
Business Reply Mail Processing Centre 969 Eastern                     16   
Rathnelly                                                             15   
Deer Park                                                             15   
South Hill                                                            15   
South Niagara                                                         15   
Harbourfront West                                                     15   
Railway Lands                                                         15   
Forest Hill SE                                                        15   
CN Tower                                                              15   
Summerhill West                                                       15   
Bathurst Quay                                                         15   
King and Spadina                                                      15   
Island airport                                                        15   
Roncesvalles                                                          14   
Parkdale               

## How many unique venues?

In [309]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 232 uniques categories.


## Analyzing the neighbourhoods

In [310]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop     ...       Theme Restaurant  Thrift / Vintage Store  \
0             0     ...                      0                       0   
1             0     ...                      0                       0   
2             0     ...                      0                       0   
3             0     ...                      0                       0   
4             0     ...                      0                       0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Wings Joint  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 232 columns]

In [311]:
toronto_onehot.shape

(3206, 232)

## neighbourhood and mean of venues

In [312]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                            Adelaide     0.000000   
1                                       Bathurst Quay     0.000000   
2                                         Berczy Park     0.000000   
3                                            Brockton     0.045455   
4   Business Reply Mail Processing Centre 969 Eastern     0.000000   
5                                            CN Tower     0.000000   
6                                         Cabbagetown     0.000000   
7                                  Central Bay Street     0.011905   
8                                           Chinatown     0.000000   
9                                            Christie     0.000000   
10                               Church and Wellesley     0.012048   
11                                     Commerce Court     0.000000   
12                                         Davisville     0.000000   
13                                   Davisville North     0.000000   
14                                          Deer Park     0.000000   
15                                    Design Exchange     0.000000   
16                                 Dovercourt Village     0.000000   
17                                           Dufferin     0.000000   
18                                   Exhibition Place     0.045455   
19                               First Canadian Place     0.000000   
20                                  Forest Hill North     0.000000   
21                                     Forest Hill SE     0.000000   
22                                   Forest Hill West     0.000000   
23                                    Garden District     0.000000   
24                                        Grange Park     0.000000   
25                                            Harbord     0.000000   
26                                       Harbourfront     0.000000   
27                                  Harbourfront East     0.000000   
28                                  Harbourfront West     0.000000   
29                                          High Park     0.000000   
..                                                ...          ...   
42                                      Railway Lands     0.000000   
43                                          Rathnelly     0.000000   
44                                           Richmond     0.000000   
45                                          Riverdale     0.024390   
46                                       Roncesvalles     0.000000   
47                                           Rosedale     0.000000   
48                                           Roselawn     0.000000   
49                                          Runnymede     0.000000   
50                                            Ryerson     0.000000   
51                                         South Hill     0.000000   
52                                      South Niagara     0.000000   
53                                     St. James Town     0.000000   
54                    Stn A PO Boxes 25 The Esplanade     0.000000   
55                                    Studio District     0.026316   
56                                    Summerhill East     0.000000   
57                                    Summerhill West     0.000000   
58                                            Swansea     0.000000   
59                                          The Annex     0.000000   
60                                        The Beaches     0.000000   
61                                   The Beaches West     0.000000   
62                                  The Danforth West     0.024390   
63                                 The Junction South     0.000000   
64                            Toronto Dominion Centre     0.000000   
65                                    Toronto Islands     0.000000   
66                                            Trinity     0.017544   
67                                   Underground ci

In [313]:
toronto_grouped.shape

(72, 232)

## Finding the top 5 most common venues

In [314]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3  Thai Restaurant  0.04
4           Bakery  0.03


----Bathurst Quay----
              venue  freq
0    Airport Lounge  0.13
1   Airport Service  0.13
2  Airport Terminal  0.13
3   Harbor / Marina  0.07
4               Bar  0.07


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1          Steakhouse  0.04
2                Café  0.04
3            Beer Bar  0.04
4  Seafood Restaurant  0.04


----Brockton----
            venue  freq
0     Coffee Shop  0.09
1            Café  0.09
2  Breakfast Spot  0.09
3     Yoga Studio  0.05
4    Intersection  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0          Smoke Shop  0.06
1       Garden Center  0.06
2          Skate Park  0.06
3      Farmers Market  0.06
4  Light Rail Station  0.06


----CN Tower----
              venue  freq
0    Airport Lounge  0.13
1

4  Italian Restaurant  0.03


----Stn A PO Boxes 25 The Esplanade----
                venue  freq
0         Coffee Shop  0.10
1          Restaurant  0.04
2                Café  0.04
3  Italian Restaurant  0.03
4  Seafood Restaurant  0.03


----Studio District----
                 venue  freq
0                 Café  0.11
1          Coffee Shop  0.08
2               Bakery  0.05
3  American Restaurant  0.05
4   Italian Restaurant  0.05


----Summerhill East----
                 venue  freq
0           Restaurant  0.33
1           Playground  0.33
2         Intersection  0.33
3  Monument / Landmark  0.00
4               Lounge  0.00


----Summerhill West----
         venue  freq
0          Pub  0.13
1  Coffee Shop  0.13
2  Pizza Place  0.07
3   Sports Bar  0.07
4   Restaurant  0.07


----Swansea----
                venue  freq
0                Café  0.08
1         Coffee Shop  0.08
2  Italian Restaurant  0.06
3    Sushi Restaurant  0.06
4           Gastropub  0.03


----The Annex----
    

## Putting it into a dataframe

In [458]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

## Create new dataframe with top 10 venues

In [459]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                           Adelaide           Coffee Shop   
1                                      Bathurst Quay       Airport Service   
2                                        Berczy Park           Coffee Shop   
3                                           Brockton        Breakfast Spot   
4  Business Reply Mail Processing Centre 969 Eastern            Skate Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1      Airport Terminal        Airport Lounge      Sculpture Garden   
2              Beer Bar            Steakhouse                Bakery   
3           Coffee Shop                  Café           Yoga Studio   
4         Burrito Place      Recording Studio  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Asian Restaurant           Salad Place            Restaurant   
1   Rental Car Location       Harbor / Marina         Boat or Ferry   
2        Farmers Market          Cocktail Bar    Seafood Restaurant   
3             Pet Store            Restaurant    Italian Restaurant   
4         Auto Workshop        Farmers Market                   Spa   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar                Bakery           Burger Joint  
1                   Bar          Airport Gate     Airport Food Court  
2           Cheese Shop                  Café         Breakfast Spot  
3          Intersection         Burrito Place                    Bar  
4           Pizza Place            Restaurant             Smoke Shop

## Now lets cluster the neighbourhoods - 3 clusters (Ran 5 clusters in the beginning but it was overlap with some neighbourhood and venues. 3 seems to be the right # of cluster for this)

In [460]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 0, 1, 1, 1, 0, 1, 1, 1, 1], dtype=int32)

In [461]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                           Adelaide           Coffee Shop   
1                                      Bathurst Quay       Airport Service   
2                                        Berczy Park           Coffee Shop   
3                                           Brockton        Breakfast Spot   
4  Business Reply Mail Processing Centre 969 Eastern            Skate Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1      Airport Terminal        Airport Lounge      Sculpture Garden   
2              Beer Bar            Steakhouse                Bakery   
3           Coffee Shop                  Café           Yoga Studio   
4         Burrito Place      Recording Studio  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0      Asian Restaurant           Salad Place            Restaurant   
1   Rental Car Location       Harbor / Marina         Boat or Ferry   
2        Farmers Market          Cocktail Bar    Seafood Restaurant   
3             Pet Store            Restaurant    Italian Restaurant   
4         Auto Workshop        Farmers Market                   Spa   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Bar                Bakery           Burger Joint  
1                   Bar          Airport Gate     Airport Food Court  
2           Cheese Shop                  Café         Breakfast Spot  
3          Intersection         Burrito Place                    Bar  
4           Pizza Place            Restaurant             Smoke Shop

## Creating new dataframe with cluster labels as well as top 10 venues

In [462]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df_toronly

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.tail() # check the last columns!

Postcode           Borough  \
69      M6R      West Toronto   
70      M6S      West Toronto   
71      M6S      West Toronto   
72      M7Y      East Toronto   
73      M9A  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
69                                       Roncesvalles  43.648960 -79.456325   
70                                          Runnymede  43.651571 -79.484450   
71                                            Swansea  43.651571 -79.484450   
72  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   
73                                       Queen's Park  43.667856 -79.532242   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
69             1.0             Gift Shop                  Bank   
70             1.0           Coffee Shop                  Café   
71             1.0           Coffee Shop                  Café   
72             1.0            Skate Park         Burrito Place   
73             NaN                   NaN                   NaN   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
69    Italian Restaurant  Eastern European Restaurant               Dog Run   
70    Italian Restaurant             Sushi Restaurant          Dessert Shop   
71    Italian Restaurant             Sushi Restaurant          Dessert Shop   
72      Recording Studio         Fast Food Restaurant         Auto Workshop   
73                   NaN                          NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
69          Dessert Shop      Cuban Restaurant         Movie Theater   
70             Bookstore        Sandwich Place                   Pub   
71             Bookstore        Sandwich Place                   Pub   
72        Farmers Market                   Spa           Pizza Place   
73                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
69            Restaurant         Breakfast Spot  
70         Burrito Place                   Park  
71         Burrito Place                   Park  
72            Restaurant             Smoke Shop  
73                   NaN                    NaN

## Dropping last row due to NaN

In [463]:
toronto_merged.drop(toronto_merged.tail(1).index,inplace=True) # drop last n rows

#converting cluster to int to avoid problems with tuple index when visualizing at next step:
toronto_merged=toronto_merged.astype({'Cluster Labels':'Int32'})

## Visualizing the resulting clusters

In [464]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

## Now proceading to examine the clusters

## Cluster 1: "The adventurer" - Airport in Downtown Toronto area. 

In [466]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]



Borough  Cluster Labels 1st Most Common Venue  \
48  Downtown Toronto               0       Airport Service   
49  Downtown Toronto               0       Airport Service   
50  Downtown Toronto               0       Airport Service   
51  Downtown Toronto               0       Airport Service   
52  Downtown Toronto               0       Airport Service   
53  Downtown Toronto               0       Airport Service   
54  Downtown Toronto               0       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
48      Airport Terminal        Airport Lounge      Sculpture Garden   
49      Airport Terminal        Airport Lounge      Sculpture Garden   
50      Airport Terminal        Airport Lounge      Sculpture Garden   
51      Airport Terminal        Airport Lounge      Sculpture Garden   
52      Airport Terminal        Airport Lounge      Sculpture Garden   
53      Airport Terminal        Airport Lounge      Sculpture Garden   
54      Airport Terminal        Airport Lounge      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
48   Rental Car Location       Harbor / Marina         Boat or Ferry   
49   Rental Car Location       Harbor / Marina         Boat or Ferry   
50   Rental Car Location       Harbor / Marina         Boat or Ferry   
51   Rental Car Location       Harbor / Marina         Boat or Ferry   
52   Rental Car Location       Harbor / Marina         Boat or Ferry   
53   Rental Car Location       Harbor / Marina         Boat or Ferry   
54   Rental Car Location       Harbor / Marina         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
48                   Bar          Airport Gate     Airport Food Court  
49                   Bar          Airport Gate     Airport Food Court  
50                   Bar          Airport Gate     Airport Food Court  
51                   Bar          Airport Gate     Airport Food Court  
52                   Bar          Airport Gate     Airport Food Court  
53                   Bar          Airport Gate     Airport Food Court  
54                   Bar          Airport Gate     Airport Food Court

## Cluster 2: "The Hangout Area" - Restaurants, bars, hotels, etc. just go out. In the area of East, Central and West Toronto

In [467]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
0      East Toronto               1  Other Great Outdoors   
1      East Toronto               1      Greek Restaurant   
2      East Toronto               1      Greek Restaurant   
3      East Toronto               1        Sandwich Place   
4      East Toronto               1        Sandwich Place   
5      East Toronto               1                  Café   
6   Central Toronto               1                  Park   
7   Central Toronto               1                 Hotel   
8   Central Toronto               1        Clothing Store   
9   Central Toronto               1           Pizza Place   
..              ...             ...                   ...   
63     West Toronto               1        Breakfast Spot   
64     West Toronto               1        Breakfast Spot   
65     West Toronto               1        Breakfast Spot   
66     West Toronto               1    Mexican Restaurant   
67     West Toronto               1    Mexican Restaurant   
68     West Toronto               1             Gift Shop   
69     West Toronto               1             Gift Shop   
70     West Toronto               1           Coffee Shop   
71     West Toronto               1           Coffee Shop   
72     East Toronto               1            Skate Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                    Pub                 Trail            Health Food Store   
1            Coffee Shop    Italian Restaurant               Ice Cream Shop   
2            Coffee Shop    Italian Restaurant               Ice Cream Shop   
3                   Park      Sushi Restaurant         Fast Food Restaurant   
4                   Park      Sushi Restaurant         Fast Food Restaurant   
5            Coffee Shop                Bakery           Italian Restaurant   
6   Gym / Fitness Center              Bus Line                  Swim School   
7            Pizza Place        Breakfast Spot               Sandwich Place   
8            Coffee Shop   Sporting Goods Shop                         Park   
9           Dessert Shop        Sandwich Place           Italian Restaurant   
..                   ...                   ...                          ...   
63           Coffee Shop                  Café                  Yoga Studio   
64           Coffee Shop                  Café                  Yoga Studio   
65           Coffee Shop                  Café                  Yoga Studio   
66                  Café       Thai Restaurant                          Bar   
67                  Café       Thai Restaurant                          Bar   
68                  Bank    Italian Restaurant  Eastern European Restaurant   
69                  Bank    Italian Restaurant  Eastern European Restaurant   
70                  Café    Italian Restaurant             Sushi Restaurant   
71                  Café    Italian Restaurant             Sushi Restaurant   
72         Burrito Place      Recording Studio         Fast Food Restaurant   

     5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0              Wings Joint                        Diner         Deli / Bodega   
1   Furniture / Home Store                   Restaurant             Bookstore   
2   Furniture / Home Store                   Restaurant             Bookstore   
3               Steakhouse               Ice Cream Shop     Fish & Chips Shop   
4               Steakhouse               Ice Cream Shop     Fish & Chips Shop   
5      American Restaurant                  Yoga Studio             Bookstore   
6             Dance Studio  Eastern European Restaurant   Dumpling Restaurant   
7        Food & Drink Shop                         Park                   Gym   
8       Chinese Restaurant               Cosmetics Shop    Mexican Restaurant   
9                      Gym                         Café      Sushi Restaurant   
..                     ...             

## Cluster 3: "The Nature Heaven area" - Known for dominance of parks, trails and restaurants. IN the Central and Downtown Toronto Area

In [468]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               2                  Park   
38   Central Toronto               2                  Park   
39   Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                 Trail            Playground      Cuban Restaurant   
38         Jewelry Store                 Trail      Sushi Restaurant   
39         Jewelry Store                 Trail      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17   Dumpling Restaurant            Donut Shop      Doner Restaurant   
38          Cupcake Shop   Dumpling Restaurant            Donut Shop   
39          Cupcake Shop   Dumpling Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17               Dog Run        Discount Store                  Diner  
38      Doner Restaurant               Dog Run         Discount Store  
39      Doner Restaurant               Dog Run         Discount Store